### ISSUE TITLE
Implement Linear Algebra Modifiers

### ISSUE THREAD

**Issue by ibaned** [gahansen/Albany#105]
This issue actually encompasses many other things including #68, #77, and #13. It is a breakaway from trilinos/Trilinos#1204. We'd like to have the option to make several different kinds of modifications to the linear algebra structures:

1. For each Dirichlet row, zero out all but the diagonal, set the RHS entry to the diagonal entry times the DBC value. This is different from what we do now, which is a one on the diagonal. The difference was suggested by @bgranzow . We could also keep an option to do either.

2. For each Dirichlet column, subtract the column times the DBC value from the RHS, then zero out the non-diagonal column terms. This actually doesn't change the Residual.

3. For each Dirichlet global element, remove it from the linear system. This is basically a reconstruction of the Jacobian and Residual structures, as well as the Solution vector.

4. For each Dirichlet global element, set the value directly in the original Solution vector.

5. Scale all rows of the linear system by some statistic of the row

Each of these stages could be on or off, for example it some of the earlier modifications aren't necessary if the constrained elements will just be removed from the system altogether.

The first obstacle to this has been determining what state the non-linear solver is in from a Thyra::ModelEvaluator level (see #77), but it seems from trilinos/Trilinos#1204 that the first thing we should try is using a NOX PrePostOperator to simply inform Albany of non-linear solver state.

**Comment by ibaned** [comment]
@bgranzow suggested some notation to help prevent confusion when dealing with the different linear algebra objects. The linear system will be denoted:
```
r = (A * x) - b 
```
Where `r` is the linear residual, `A` is the linear jacobian, `x` is the linear solution vector, `b` is the linear force vector which I call the RHS above.
The nonlinear system for the Newton method is:
```
R = (A * delta_x) + r
```
Where `R` is the "nonlinear residual", and `delta_x` is the Newton step direction vector.

**Comment by bartgol** [comment]
@ibaned I'm not sure what you mean by "nonlinear system" and "nonlinear residual". Isn't the A in the second equation the Jacobian?

**Comment by ibaned** [comment]
Correct, when executing Newton's method that "nonlinear system" is really the only linear system that is solved. Its just useful to establish how it is different from the classic `A*x=b` that is presented in finite element discussions for linear physics problems. I'll edit my comment to make the names better.

**Comment by lxmota** [comment]
I would like to get from NOX the solution vector at the beginning of the solution procedure (`x_init`) and at the end once it converges to the solution (`x_final`). Is that possible through the PrePostOperator?

For the Schwarz alternating method, the convergence criterion is the following:

`\sum_{i=1}^{N}{(x_final^i-x_init^i)^2/(x_final^i)^2}`

for N subdomains. This all happens in a Schwarz loop here:

https://github.com/gahansen/Albany/blob/master/src/LCM/Schwarz_Alternating.cpp#L390

where the `solver` for each subdomain is called here:

https://github.com/gahansen/Albany/blob/master/src/LCM/Schwarz_Alternating.cpp#L414

where the `solver` object for each subdomain is a `Piro::NoxSolver`.

I don't know to get the solution vectors from the `Piro::NoxSolver` and that is preventing me from making progress on the Schwarz implementation. Any suggestions?

**Comment by ibaned** [comment]
This is the type of thing we should be able to do inside Albany once we have this context information. When the PrePostOperator informs us that the non-linear solve has started, save the solution vectors (which we can start tracking more carefully, from calls to the ModelEvaluator). Then when the PrePostOperator tells us the solve is done, compare the current solution to the saved one. I know this is vague but in general that is what I think we should do.

**Comment by lxmota** [comment]
Sounds good. I'll give it a try.
